### ISSUE TITLE
Evaluate on test set during training?

### ISSUE THREAD

**Issue by Karol-G** [MIC-DKFZ/nnUNet#307]
Hi Fabian,

I am currently training on a private dataset. The dataset is split into a train and test set.  As far as I understand the nnUNet uses 5-fold cross validation on the train set to estimate its performance during training. 
Is there away to estimate the performance on the test set instead? 
I finished one training run already and evaluted some of the checkpoints during training on the test set. Some of these achieved dice scores of +0.6. On the other hand the "best" and latest checkpoints saved by the nnUNet after the training finished merely achieved a dice score of about 0.5 on the test set.

Is this funtionality implemented somewhere in the nnUNet? 
Alternatively could you guide me to the .py files or sections where I would need to implement this. The framework is quite big and I am still not certain what happens where, sorry for this :/

Best
Karol

**Comment by FabianIsensee** [comment]
Hi Karol,
this is **absolutely not** what test sets are for and I **very strongly** recommend not doing that. Your message sounds like you want to overfit to the test set. Do not do that! Put your test set away and don't touch it.
If you observe wildly varying performance then you should probably consider increasing the size and diversity of the training set.
Best,
Fabian

**Comment by Karol-G** [comment]
Hi Fabian,

Probably should have been a bit more precise ^^
I only want to monitor the performance on the test set, not overfit to it (No hyperparameter optimization, early stopping, choosing best checkpoint based on test set, etc).
This is more out of curiosity and to get a better feeling of the framework and my data.

I have already implemented this now as I assume this was not implemented.

Best
Karol

**Comment by FabianIsensee** [comment]
Hi Karol,
the intent does not matter. If you look at the test set at this stage you are breaking best practices. If you see it works better on some epoch than it does towards the end then you will be inclined to do epoch selection based on that and that is really not what one should do.
If you want to get a feeling for the framework please use a validation set. You can just run a single fold and let nnunet split your data into train and val.
Best,
Fabian
